File: ruby/core/lib/mullet/html/template_builder.rb
require 'mullet/html/attributes'
require 'mullet/html/command'
require 'mullet/html/element'
require 'mullet/html/element_renderer'
require 'mullet/html/for_element_renderer'
require 'mullet/html/if_element_renderer'
require 'mullet/html/parser/default_handler'
require 'mullet/html/parser/simple_parser'
require 'mullet/html/remove_mode'
require 'mullet/html/static_text_renderer'
require 'mullet/html/template'
require 'mullet/html/unless_element_renderer'
require 'mullet/template_error'
require 'set'

module Mullet; module HTML

  # Handles SAX events to build a template.
  class TemplateBuilder < Parser::DefaultHandler
    include Command

    COMMANDS = Set[
        ACTION,
        ALT,
        ALT_MESSAGE,
        ATTR,
        ATTR_MESSAGE,
        ESCAPE_XML,
        FOR,
        HREF,
        IF,
        INCLUDE,
        REMOVE,
        SRC,
        TEXT,
        TEXT_MESSAGE,
        TITLE,
        TITLE_MESSAGE,
        UNLESS,
        VALUE,
        VALUE_MESSAGE]
    START_CDATA = '<![CDATA['
    END_CDATA = ']]>'

    attr_reader :template

    # Constructor
    #
    # @param [TemplateLoader] loader
    #           template loader to use to load included template files
    def initialize(loader)
      @loader = loader

      # Stack of elements where this handler has seen the start tag and not yet
      # seen the end tag.
      @open_elements = []

      # stack of current containers to add renderers to
      @containers = []
    
      @static_text = ''
      @template = nil
    end

    # Adds renderer to current container.
    #
    # @param [#render] renderer
    #           renderer to add
    def add_child(renderer)
      @containers.last().add_child(renderer)
    end

    # Deletes renderer from current container.
    #
    # @param [#render] renderer
    #           renderer to delete
    def delete_child(renderer)
      @containers.last().delete_child(renderer)
    end

    # Partitions the attributes into ordinary and command attributes.
    #
    # @param [Array] attributes
    #           input attributes
    # @param [Hash] ns
    #           hash of namespace prefix to uri mappings
    # @param [#store] ordinary_attributes
    #           hash will receive name to value mappings for ordinary attributes
    # @param [#store] command_attributes
    #           hash will receive name to value mappings for command attributes
    # @return [Boolean] true if any command attribute found
    def find_commands(attributes, ns, ordinary_attributes, command_attributes)
      found_command = false
      attributes.each do |attr|
        if attr.localname.start_with?(DATA_PREFIX)
          command_name = attr.localname.slice(DATA_PREFIX.length()..-1)
          if COMMANDS.include?(command_name)
            command_attributes.store(command_name, attr.value)
            found_command = true
          end
        elsif attr.uri == NAMESPACE_URI || attr.prefix == NAMESPACE_PREFIX
          command_name = attr.localname
          if !COMMANDS.include?(command_name)
            raise TemplateError("invalid command '#{command_name}'")
          end
          command_attributes.store(command_name, attr.value)
          found_command = true
        else
          attribute_name = [attr.prefix, attr.localname].compact().join(':')
          ordinary_attributes.store(attribute_name, attr.value)
        end
      end

      ns.each do |prefix, uri|
        if uri != NAMESPACE_URI
          attribute_name = ['xmlns', prefix].compact().join(':')
          ordinary_attributes.store(attribute_name, uri)
        end
      end

      return found_command
    end

    def append_static_text(data)
      @static_text << data
    end

    def end_static_text()
      if !@static_text.empty?()
        add_child(StaticTextRenderer.new(@static_text))
        @static_text = ''
      end
    end

    # Marks the most deeply nested open element as having content.
    def set_has_content()
      if !@open_elements.empty?()
        @open_elements.last().has_content = true
      end
    end

    def start_document()
      @template = Template.new()
      @containers.push(@template)
    end

    def end_document()
      end_static_text()
    end

    def doctype(name, public_id, system_id)
      text = "<!DOCTYPE #{name}"

      if public_id
        text << %( PUBLIC "#{public_id}")
      end

      if system_id
        text << %( "#{system_id}")
      end

      text << '>'
      append_static_text(text)
    end

    def create_element_renderer(element, command_attributes)
      renderer = nil

      variable_name = command_attributes.fetch(IF, nil)
      if variable_name != nil
        renderer = IfElementRenderer.new(element, variable_name)
      end

      if renderer == nil
        variable_name = command_attributes.fetch(UNLESS, nil)
        if variable_name != nil
          renderer = UnlessElementRenderer.new(element, variable_name)
        end
      end

      if renderer == nil
        variable_name = command_attributes.fetch(FOR, nil)
        if variable_name != nil
          renderer = ForElementRenderer.new(element, variable_name)
        end
      end

      if renderer == nil
        renderer = ElementRenderer.new(element)
      end

      renderer.configure_commands(command_attributes, @loader)
      return renderer
    end

    def start_element_namespace(name, attributes, prefix, uri, ns)
      set_has_content()

      ordinary_attributes = Attributes.new()
      command_attributes = Attributes.new()
      found_command = find_commands(
          attributes, ns, ordinary_attributes, command_attributes)

      qualified_name = [prefix, name].compact().join(':')
      element = Element.new(qualified_name, ordinary_attributes)
      @open_elements.push(element)

      if found_command
        end_static_text()

        element.has_command = true
        renderer = create_element_renderer(element, command_attributes)
        add_child(renderer)

        @containers.push(renderer)
      else
        append_static_text(element.render_start_tag(element.attributes))
      end
    end

    def configure_remove_command(element, renderer)
      case renderer.remove_mode
      when RemoveMode::TAG
        if !renderer.has_command && !renderer.has_dynamic_content
          # Discard tag, but preserve the children.
          delete_child(renderer)
          renderer.children.each do |child|
            add_child(child)
          end
        end
      when RemoveMode::CONTENT
        if !renderer.has_command
          # Discard children. Statically render the tag.
          delete_child(renderer)

          append_static_text(element.render_start_tag(element.attributes))
          append_static_text(element.render_end_tag())
        end
      when RemoveMode::ELEMENT
        # Discard element and all its content.
        delete_child(renderer)
      end
    end

    def end_element_namespace(name, prefix, uri)
      element = @open_elements.pop()
      if element.has_command
        end_static_text()

        renderer = @containers.pop()
        if renderer.has_dynamic_content
          # Discard children because the content will be replaced at render
          # time.
          renderer.clear_children()
        end

        configure_remove_command(element, renderer)
      elsif uri != Parser::SimpleParser::IMPLICIT_END_TAG_NS_URI
        append_static_text(element.render_end_tag())
      end
    end

    def characters(data)
      set_has_content()
      append_static_text(data)
    end

    def cdata_block(data)
      append_static_text(START_CDATA)
      characters(data)
      append_static_text(END_CDATA)
    end

    def comment(data)
      append_static_text('<!--')
      characters(data)
      append_static_text('-->')
    end

    def processing_instruction(data)
      append_static_text('<?')
      characters(data)
      append_static_text('?>')
    end
  end

end; end

File: java/core/src/main/java/com/github/pukkaone/mullet/html/TemplateBuilder.java
/*
Copyright (c) 2011, Chin Huang
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.github.pukkaone.mullet.html;

import com.github.pukkaone.mullet.Container;
import com.github.pukkaone.mullet.Renderer;
import com.github.pukkaone.mullet.TemplateException;
import com.github.pukkaone.mullet.html.parser.SimpleParser;
import java.io.IOException;
import java.io.StringReader;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.ext.DefaultHandler2;

/**
 * Handles SAX events to build a template.
 */
class TemplateBuilder extends DefaultHandler2 {
    private static final HashSet<String> COMMANDS =
            new HashSet<String>(Arrays.asList(
                    Command.ACTION,
                    Command.ALT,
                    Command.ALT_MESSAGE,
                    Command.ATTR,
                    Command.ATTR_MESSAGE,
                    Command.ESCAPE_XML,
                    Command.FOR,
                    Command.HREF,
                    Command.IF,
                    Command.INCLUDE,
                    Command.REMOVE,
                    Command.SRC,
                    Command.TEXT,
                    Command.TEXT_MESSAGE,
                    Command.TITLE,
                    Command.TITLE_MESSAGE,
                    Command.UNLESS,
                    Command.VALUE,
                    Command.VALUE_MESSAGE));
    private static final char[] START_CDATA = "<![CDATA[".toCharArray();
    private static final char[] END_CDATA = "]]>".toCharArray();

    private TemplateLoader loader;

    // This is a stack of elements where this handler has seen the start tag and
    // not yet seen the end tag.
    private LinkedList<Element> openElements = new LinkedList<Element>();

    // stack of current containers to add renderers to
    private LinkedList<Container> containers = new LinkedList<Container>();

    private StringBuilder staticText = new StringBuilder();
    private Template template;

    TemplateBuilder(TemplateLoader loader) {
        this.loader = loader;
    }

    private void addChild(Renderer renderer) {
        containers.getFirst().addChild(renderer);
    }

    private void removeChild(Renderer renderer) {
        containers.getFirst().removeChild(renderer);
    }

    private boolean findCommands(
            org.xml.sax.Attributes inputAttributes,
            Attributes ordinaryAttributes,
            Attributes commandAttributes)
    {
        boolean foundCommand = false;
        for (int i = 0; i < inputAttributes.getLength(); ++i) {
            String value = inputAttributes.getValue(i);

            if (Command.NAMESPACE_URI.equals(inputAttributes.getURI(i))
             || inputAttributes.getQName(i).startsWith(Command.PREFIX))
            {
                String commandName = inputAttributes.getLocalName(i);
                if (!COMMANDS.contains(commandName)) {
                    throw new TemplateException(
                            "invalid command '" + commandName + "'");
                }
                commandAttributes.put(commandName, value);
                foundCommand = true;
            } else {
                String attributeName = inputAttributes.getQName(i);
                if (attributeName.length() == 0) {
                    attributeName = inputAttributes.getLocalName(i);
                }
                ordinaryAttributes.put(attributeName, value);
            }
        }
        return foundCommand;
    }

    private void appendStaticText(char[] ch, int start, int length) {
        staticText.append(ch, start, length);
    }

    private void appendStaticText(String text) {
        staticText.append(text);
    }

    private void endStaticText() {
        if (staticText.length() > 0) {
            addChild(new StaticTextRenderer(staticText.toString()));
            staticText.delete(0, staticText.length());
        }
    }

    // Marks the most deeply nested open element as having content.
    private void setHasContent() {
        if (!openElements.isEmpty()) {
            openElements.getFirst().hasContent = true;
        }
    }

    @Override
    public InputSource resolveEntity(
            String name,
            String publicId,
            String baseURI,
            String systemId)
        throws SAXException, IOException
    {
        // Return empty input for all entities, otherwise the parser by default
        // tries to read the DTD from the Internet if there is a doctype
        // declaration with a system ID.
        return new InputSource(new StringReader(""));
    }

    @Override
    public void startDocument() throws SAXException {
        template = new Template();
        containers.addFirst(template);
    }

    @Override
    public void endDocument() throws SAXException {
        endStaticText();
    }

    @Override
    public void startDTD(String name, String publicId, String systemId)
         throws SAXException
    {
        StringBuilder doctype = new StringBuilder();
        doctype.append("<!DOCTYPE ").append(name);
        if (publicId != null) {
            doctype.append(" PUBLIC \"").append(publicId).append('"');
        }
        if (systemId != null) {
            doctype.append(" \"").append(systemId).append('"');
        }
        doctype.append(">");

        appendStaticText(doctype.toString());
    }

    private ElementRenderer createElementRenderer(
            Element element, Attributes commandAttributes)
    {
        ElementRenderer renderer = null;

        String value = commandAttributes.get(Command.IF);
        if (value != null) {
            renderer = new IfElementRenderer(element, value);
        }

        if (renderer == null) {
            value = commandAttributes.get(Command.UNLESS);
            if (value != null) {
                renderer = new UnlessElementRenderer(element, value);
            }
        }

        if (renderer == null) {
            value = commandAttributes.get(Command.FOR);
            if (value != null) {
                renderer = new ForElementRenderer(element, value);
            }
        }

        if (renderer == null) {
            renderer = new ElementRenderer(element);
        }

        if (!(renderer instanceof IncludeElementRenderer)) {
            // The include command replaces the entire element, so we don't have
            // handle other commands in the element.
            renderer.configureCommands(commandAttributes, loader);
        }
        return renderer;
    }

    @Override
    public void startElement(
            String namespaceUri,
            String localName,
            String qualifiedName,
            org.xml.sax.Attributes attributes)
        throws SAXException
    {
        setHasContent();

        Attributes ordinaryAttributes = new Attributes();
        Attributes commandAttributes = new Attributes();
        boolean foundCommand = findCommands(
                attributes, ordinaryAttributes, commandAttributes);

        Element element = new Element(
                localName, qualifiedName, ordinaryAttributes);
        openElements.addFirst(element);

        if (foundCommand) {
            endStaticText();

            element.hasCommand = true;
            ElementRenderer renderer = createElementRenderer(
                    element, commandAttributes);
            addChild(renderer);

            containers.addFirst(renderer);
        } else {
            appendStaticText(element.renderStartTag(element.attributes));
        }
    }

    private void configureRemoveCommand(
            Element element, ElementRenderer renderer)
    {
        if (renderer.removeMode == null) {
            return;
        }

        switch (renderer.removeMode) {
        case ELEMENT:
            // Discard element and all its content.
            removeChild(renderer);
            break;

        case TAG:
            if (!renderer.hasCommand() && !renderer.hasDynamicContent()) {
                // Discard tag, but preserve the children.
                removeChild(renderer);
                for (Renderer child : renderer.getChildren()) {
                    addChild(child);
                }
            }
            break;

        case CONTENT:
            if (!renderer.hasCommand()) {
                // Discard children.  Statically render the tag.
                removeChild(renderer);

                appendStaticText(element.renderStartTag(element.attributes));
                appendStaticText(element.renderEndTag());
            }
            break;
        }
    }

    @Override
    public void endElement(
            String namespaceUri, String localName, String qualifiedName)
        throws SAXException
    {
        Element element = openElements.removeFirst();
        if (element.hasCommand) {
            endStaticText();

            ElementRenderer renderer =
                    (ElementRenderer) containers.removeFirst();
            if (renderer.hasDynamicContent()) {
                // Discard children because the content will be replaced at
                // render time.
                renderer.clearChildren();
            }

            configureRemoveCommand(element, renderer);
        } else if (!SimpleParser.IMPLICIT_END_TAG_NS_URI.equals(namespaceUri)) {
            appendStaticText(element.renderEndTag());
        }
    }

    @Override
    public void characters(char[] ch, int start, int length)
        throws SAXException
    {
        setHasContent();

        appendStaticText(ch, start, length);
    }

    @Override
    public void startCDATA() throws SAXException {
        characters(START_CDATA, 0, START_CDATA.length);
    }

    @Override
    public void endCDATA() throws SAXException {
        characters(END_CDATA, 0, END_CDATA.length);
    }

    @Override
    public void ignorableWhitespace(char[] ch, int start, int length)
        throws SAXException
    {
        characters(ch, start, length);
    }

    @Override
    public void comment(char[] ch, int start, int length)
        throws SAXException
    {
        appendStaticText("<!--");
        characters(ch, start, length);
        appendStaticText("-->");
    }

    @Override
    public void processingInstruction(String target, String data)
        throws SAXException
    {
        setHasContent();

        appendStaticText("<?");
        appendStaticText(target);
        appendStaticText(" ");
        appendStaticText(data);
        appendStaticText("?>");
    }

    Template getTemplate() {
        return template;
    }
}
